src/Network/Transport/TCP.hs

```diff
@@ -1119,17 +1119,29 @@ handleIncomingMessages (ourEndPoint, theirEndPoint) = do
             forM_ (Set.elems $ vst ^. remoteIncoming) $
               qdiscEnqueue' ourQueue theirAddr . ConnectionClosed . connId
             let vst' = remoteIncoming ^= Set.empty $ vst
-            -- If we still have outgoing connections then we ignore the
-            -- CloseSocket request (we sent a ConnectionCreated message to the
-            -- remote endpoint, but it did not receive it before sending the
-            -- CloseSocket request). Similarly, if lastReceivedId < lastSentId
-            -- then we sent a ConnectionCreated *AND* a ConnectionClosed
-            -- message to the remote endpoint, *both of which* it did not yet
-            -- receive before sending the CloseSocket request.
-            if vst' ^. remoteOutgoing > 0 || lastReceivedId < lastSentId vst
+            -- The peer sends the connection id of the last connection which
+            -- they accepted from us.
+            --
+            -- If it's not the same as the id of the last connection that we
+            -- have made to them (assuming we haven't cycled through all
+            -- identifiers so fast) then they hadn't seen the request before
+            -- they tried to close the socket. In that case, we don't close the
+            -- socket. They'll see our in-flight connection request and then
+            -- abandon their attempt to close the socket.
+            --
+            -- If it *is* the same then we can close the socket. However, if
+            -- our remoteOutgoing is nonzero, then the peer has made an error:
+            -- we haven't closed all of our lightweight connections to them, so
+            -- they should not send us CloseSocket.
+            -- In this case, we must enqueue EventConnectionLost.
+            if lastReceivedId /= lastSentId vst
               then
                 return (RemoteEndPointValid vst', Nothing)
               else do
+                when (vst' ^. remoteOutgoing > 0) $ do
```

Comment:
I think it only matters in case `lastReceived = lastSendId vst`. If `lastReceived /= lastSendId vst` then we know we had at least one new connection in flight when the peer sent `CloseSocket`, so the number of `remoteOutgoing` connections is irrelevant.